File: go/jump_game_iii.go
package leetcode

// TC: O(n)
// SC: O(n)
// optimized to not use auxiliary state to track visited nodes
func canReach(arr []int, start int) bool {
	if arr[start] == 0 {
		return true
	}

	N := len(arr)
	stack := make([][]int, 0)
	stack = append(stack, []int{start, arr[start]})
	arr[start] = -1

	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		i, d := node[0], node[1]

		forward, backward := i+d, i-d
		if forward < N && arr[forward] != -1 {
			if arr[forward] == 0 {
				return true
			}
			stack = append(stack, []int{forward, arr[forward]})
			arr[forward] = -1
		}
		if backward >= 0 && arr[backward] != -1 {
			if arr[backward] == 0 {
				return true
			}
			stack = append(stack, []int{backward, arr[backward]})
			arr[backward] = -1
		}
	}
	return false
}

// TC: O(n)
// SC: O(n)
func canReach2(arr []int, start int) bool {
	if arr[start] == 0 {
		return true
	}

	N := len(arr)
	dp := make([]bool, N)
	stack := make([][]int, 0, N)
	stack = append(stack, []int{start, arr[start]})
	dp[start] = true

	for len(stack) > 0 {
		node := stack[len(stack)-1]
		i, d := node[0], node[1]
		stack = stack[:len(stack)-1]

		forward, backward := i+d, i-d
		if forward < N && dp[forward] == false {
			if arr[forward] == 0 {
				return true
			}
			stack = append(stack, []int{forward, arr[forward]})
			dp[forward] = true
		}
		if backward >= 0 && dp[backward] == false {
			if arr[backward] == 0 {
				return true
			}
			stack = append(stack, []int{backward, arr[backward]})
			dp[backward] = true
		}
	}
	return false
}

File: python/jump-game-iii.py
from typing import List

class Solution:
    # TC: O(n)
    # SC: O(n)
    # optimized to avoid creating auxiliary space to keep track of visited nodes
    def canReach(self, arr: List[int], start: int) -> bool:
        if arr[start] == 0: return True
        N = len(arr)
        stack = [(start, arr[start])]
        arr[start] = -1
        
        while stack:
            i, d = stack.pop()
            forward, backward = i+d, i-d
            
            if forward < N and arr[forward] != -1:
                if arr[forward] == 0:
                    return True
                stack.append((forward, arr[forward]))
                arr[forward] = -1
            if backward >= 0 and arr[backward] != -1:
                if arr[backward] == 0:
                    return True
                stack.append((backward, arr[backward]))
                arr[backward] = -1
        return False
        
    # TC: O(n)
    # SC: O(n)
    def canReach2(self, arr: List[int], start: int) -> bool:
        if arr[start] == 0: return True
        N = len(arr)
        dp = [False]*(N)
        dp[start] = True
        stack = [(start, arr[start])]
        
        while stack:
            i, d = stack.pop()
            if i + d < N and dp[d+i] == False:
                ni = i + d
                if arr[ni] == 0:
                    return True
                stack.append((ni, arr[ni]))
                dp[ni] = True
            if i - d >= 0 and dp[i-d] == False:
                ni = i - d
                if arr[ni] == 0:
                    return True
                stack.append((ni, arr[ni]))
                dp[ni] = True
        return False
        